File: scripts/solar/solar-common/util/system/BuffUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["7"] = 1,["8"] = 1,["9"] = 2,["10"] = 2,["11"] = 3,["12"] = 3,["13"] = 26,["14"] = 26,["15"] = 26,["17"] = 26,["18"] = 33,["19"] = 35,["20"] = 36,["21"] = 37,["22"] = 37,["24"] = 39,["26"] = 33,["27"] = 48,["28"] = 49,["29"] = 50,["30"] = 51,["33"] = 54,["34"] = 48,["35"] = 62,["36"] = 63,["37"] = 64,["38"] = 65,["41"] = 68,["42"] = 62,["43"] = 74,["44"] = 75,["45"] = 76,["46"] = 77,["49"] = 80,["50"] = 74,["51"] = 86,["52"] = 87,["53"] = 88,["54"] = 89,["57"] = 92,["58"] = 86,["59"] = 98,["60"] = 99,["61"] = 100,["62"] = 101,["65"] = 104,["66"] = 98,["67"] = 112,["68"] = 113,["69"] = 114,["70"] = 112,["71"] = 125,["72"] = 126,["73"] = 127,["74"] = 128,["75"] = 129,["77"] = 132,["78"] = 133,["79"] = 134,["80"] = 136,["81"] = 137,["82"] = 138,["83"] = 139,["84"] = 139,["85"] = 139,["86"] = 140,["87"] = 139,["88"] = 139,["92"] = 145,["93"] = 146,["95"] = 149,["96"] = 149,["97"] = 149,["98"] = 149,["99"] = 149,["100"] = 149,["101"] = 149,["102"] = 150,["103"] = 125,["104"] = 160,["105"] = 161,["107"] = 162,["108"] = 162,["109"] = 163,["110"] = 164,["113"] = 167,["114"] = 168,["115"] = 169,["117"] = 171,["118"] = 162,["121"] = 173,["122"] = 160,["123"] = 181,["124"] = 182,["125"] = 182,["126"] = 181,["127"] = 190,["128"] = 191,["129"] = 191,["130"] = 190,["131"] = 198,["132"] = 199,["133"] = 200,["134"] = 198,["135"] = 207,["136"] = 208,["137"] = 209,["138"] = 210,["139"] = 211,["141"] = 213,["142"] = 214,["143"] = 215,["144"] = 215,["145"] = 216,["146"] = 217,["149"] = 220,["150"] = 207,["151"] = 228,["152"] = 230,["153"] = 231,["154"] = 232,["155"] = 233,["158"] = 236,["161"] = 228,["162"] = 246,["163"] = 247,["164"] = 249,["165"] = 250,["166"] = 251,["167"] = 252,["168"] = 253,["169"] = 254,["170"] = 255,["174"] = 249,["176"] = 262,["177"] = 263,["178"] = 264,["179"] = 265,["180"] = 266,["183"] = 262,["185"] = 246,["186"] = 277,["187"] = 278,["188"] = 279,["191"] = 282,["192"] = 283,["193"] = 284,["194"] = 285,["197"] = 277,["198"] = 296,["199"] = 297,["201"] = 297,["203"] = 296,["204"] = 305,["205"] = 305,["206"] = 305,["208"] = 306,["209"] = 307,["212"] = 310,["213"] = 311,["214"] = 312,["215"] = 312,["216"] = 313,["219"] = 316,["221"] = 316,["225"] = 305,["226"] = 327,["227"] = 328,["228"] = 329,["231"] = 332,["232"] = 333,["233"] = 334,["234"] = 334,["235"] = 335,["238"] = 327,["239"] = 344,["240"] = 345,["241"] = 345,["242"] = 344,["243"] = 351,["244"] = 352,["245"] = 352,["246"] = 351,["247"] = 358,["248"] = 359,["249"] = 359,["250"] = 358,["251"] = 369,["252"] = 370,["255"] = 373,["256"] = 374,["257"] = 375,["260"] = 378,["261"] = 369,["262"] = 388,["263"] = 389,["266"] = 392,["267"] = 393,["268"] = 394,["271"] = 397,["272"] = 388,["273"] = 28});
local ____exports = {}
local ____Buff = require("solar.solar-common.tool.Buff")
local Buff = ____Buff.default
local ____DataBase = require("solar.solar-common.common.DataBase")
local DataBase = ____DataBase.default
local ____BaseUtil = require("solar.solar-common.util.BaseUtil")
local BaseUtil = ____BaseUtil.default
____exports.default = __TS__Class()
local BuffUtil = ____exports.default
BuffUtil.name = "BuffUtil"
function BuffUtil.prototype.____constructor(self)
end
function BuffUtil.registerBuffType(self, buffType)
    if DataBase:getSolarBuffType(buffType.id) == nil then
        DataBase:setSolarBuffType(buffType.id, buffType)
        local ____exports_default_buffTypes_0 = ____exports.default.buffTypes
        ____exports_default_buffTypes_0[#____exports_default_buffTypes_0 + 1] = buffType
    else
        log.errorWithTraceBack((((("不能重复注册BuffType:" .. buffType.id) .. " -> ") .. tostring(buffType.name)) .. " -> ") .. DataBase:getSolarBuffType(buffType.id).name)
    end
end
function BuffUtil.registerBuffCreatedEvent(self, buffTypeId, onCreated)
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType.onCreated then
        log.errorWithTraceBack("不能重复注册Buff事件:" .. buffTypeId)
        return
    end
    buffType.onCreated = onCreated
end
function BuffUtil.registerBuffUpdateEvent(self, buffTypeId, onUpdate)
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType.onUpdate then
        log.errorWithTraceBack("不能重复注册Buff事件:" .. buffTypeId)
        return
    end
    buffType.onUpdate = onUpdate
end
function BuffUtil.registerBuffIntervalEvent(self, buffTypeId, onInterval)
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType.onInterval then
        log.errorWithTraceBack("不能重复注册Buff事件:" .. buffTypeId)
        return
    end
    buffType.onInterval = onInterval
end
function BuffUtil.registerBuffStackCountChangeEvent(self, buffTypeId, onStackCountChange)
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType.onStackCountChange then
        log.errorWithTraceBack("不能重复注册Buff事件:" .. buffTypeId)
        return
    end
    buffType.onStackCountChange = onStackCountChange
end
function BuffUtil.registerBuffDestroyEvent(self, buffTypeId, onDestroy)
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType.onDestroy then
        log.errorWithTraceBack("不能重复注册Buff事件:" .. buffTypeId)
        return
    end
    buffType.onDestroy = onDestroy
end
function BuffUtil.getBuffType(self, buffTypeId)
    local bt = DataBase:getSolarBuffType(buffTypeId)
    return bt
end
function BuffUtil.addBuff(self, unit, buffTypeId, caster, initBuff)
    local buffType = DataBase:getSolarBuffType(buffTypeId)
    if not buffType then
        log.errorWithTraceBack("buff不存在!请先使用registerBuffType注册Buff类型! id=" .. buffTypeId)
        return nil
    end
    local oldUnitBuff = ____exports.default:getUnitBuff(unit, buffTypeId)
    if oldUnitBuff then
        if buffType.stackCountMax and buffType.stackCountMax ~= 0 then
            if buffType.stackCountMax == -1 or oldUnitBuff.stackCount < buffType.stackCountMax then
                oldUnitBuff.stackCount = oldUnitBuff.stackCount + 1
                if buffType.stackDurMode then
                    BaseUtil.runLater(
                        buffType.dur,
                        function()
                            oldUnitBuff.stackCount = oldUnitBuff.stackCount - 1
                        end
                    )
                end
            end
        end
        oldUnitBuff:refresh()
        return oldUnitBuff
    end
    local buff = __TS__New(
        Buff,
        buffType,
        unit,
        caster,
        initBuff
    )
    return buff
end
function BuffUtil.getPlayerUnitBuff(self, player, buffTypeId)
    GroupEnumUnitsOfPlayer(_sl_tempGroup3, player, nil)
    do
        local i = 0
        while i <= 1000000 do
            local unitHandle = FirstOfGroup(_sl_tempGroup3)
            if not IsHandle(unitHandle) then
                break
            end
            local buff = ____exports.default:getUnitBuff(unitHandle, buffTypeId)
            if (buff and buff.buffType.id) == buffTypeId then
                return buff
            end
            GroupRemoveUnit(_sl_tempGroup3, unitHandle)
            i = i + 1
        end
    end
    return nil
end
function BuffUtil.getUnitBuff(self, unit, buffTypeId)
    local ____opt_3 = ____exports.default:getUnitBuffs(unit)
    return ____opt_3 and ____opt_3[buffTypeId]
end
function BuffUtil.hasUnitBuff(self, unit, buffTypeId)
    local ____opt_5 = ____exports.default:getUnitBuffs(unit)
    return (____opt_5 and ____opt_5[buffTypeId]) ~= nil
end
function BuffUtil.getUnitBuffs(self, unit)
    local unitSolarData = DataBase:getUnitSolarData(unit, false)
    return unitSolarData and unitSolarData._SL_solarBuffSet
end
function BuffUtil.getUnitBuffAttributes(self, unit)
    local unitSolarData = DataBase:getUnitSolarData(unit, false)
    local solarBuffSet = unitSolarData and unitSolarData._SL_solarBuffSet
    if not solarBuffSet then
        return nil
    end
    local attributes = {}
    for solarBuffSetKey in pairs(solarBuffSet) do
        local ____opt_11 = solarBuffSet[solarBuffSetKey]
        local attribute = ____opt_11 and ____opt_11.attribute
        if attribute then
            attributes[#attributes + 1] = attribute
        end
    end
    return attributes
end
function BuffUtil.forAllBuffTypes(self, callback, buffTypeClass)
    for ____, buffType in ipairs(____exports.default.buffTypes) do
        if buffTypeClass then
            if buffTypeClass == buffType.class then
                callback(nil, buffType)
            end
        else
            callback(nil, buffType)
        end
    end
end
function BuffUtil.forAllBuffs(self, callback, buffTypeId)
    if buffTypeId then
        DataBase:forUnitSolarDatas(function(____, id, solarData)
            local _SL_solarBuffSet = solarData and solarData._SL_solarBuffSet
            if _SL_solarBuffSet then
                for tempBuffTypeId in pairs(_SL_solarBuffSet) do
                    local buff = _SL_solarBuffSet[tempBuffTypeId]
                    if (buff and buff.buffType.id) == buffTypeId then
                        callback(nil, buff)
                    end
                end
            end
        end)
    else
        DataBase:forUnitSolarDatas(function(____, id, solarData)
            local _SL_solarBuffSet = solarData and solarData._SL_solarBuffSet
            if _SL_solarBuffSet then
                for tempBuffTypeId in pairs(_SL_solarBuffSet) do
                    callback(nil, _SL_solarBuffSet[tempBuffTypeId])
                end
            end
        end)
    end
end
function BuffUtil.forUnitBuffs(self, unit, callback)
    local buffSet = ____exports.default:getUnitBuffs(unit)
    if not buffSet then
        return
    end
    for buffTypeId in pairs(buffSet) do
        local buff = buffSet[buffTypeId]
        if buff then
            callback(nil, buff)
        end
    end
end
function BuffUtil.clearUnitBuff(self, unit, buffTypeId)
    local ____opt_19 = ____exports.default:getUnitBuff(unit, buffTypeId)
    if ____opt_19 ~= nil then
        ____opt_19:destroy()
    end
end
function BuffUtil.clearUnitBuffs(self, unit, onlyDeBuff)
    if onlyDeBuff == nil then
        onlyDeBuff = false
    end
    local buffs = ____exports.default:getUnitBuffs(unit)
    if not buffs then
        return
    end
    for buffsKey in pairs(buffs) do
        if onlyDeBuff then
            local ____opt_21 = buffs[buffsKey]
            if ____opt_21 and ____opt_21.buffType.isDebuff then
                buffs[buffsKey]:destroy()
            end
        else
            local ____opt_23 = buffs[buffsKey]
            if ____opt_23 ~= nil then
                ____opt_23:destroy()
            end
        end
    end
end
function BuffUtil.clearUnitBuffsByClass(self, unit, clazz)
    local buffs = ____exports.default:getUnitBuffs(unit)
    if not buffs then
        return
    end
    for buffsKey in pairs(buffs) do
        local buff = buffs[buffsKey]
        local ____opt_25 = buff and buff.buffType
        if (____opt_25 and ____opt_25.class) == clazz then
            buff:destroy()
        end
    end
end
function BuffUtil.addAnyBuffCreatedListener(self, onBuffCreatedListener)
    local ____Buff__sl_anyBuffCreatedListeners_29 = Buff._sl_anyBuffCreatedListeners
    ____Buff__sl_anyBuffCreatedListeners_29[#____Buff__sl_anyBuffCreatedListeners_29 + 1] = onBuffCreatedListener
end
function BuffUtil.addAnyBuffStackCountChangeListener(self, onBuffStackCountChangeListener)
    local ____Buff__sl_anyBuffStackCountChangeListeners_30 = Buff._sl_anyBuffStackCountChangeListeners
    ____Buff__sl_anyBuffStackCountChangeListeners_30[#____Buff__sl_anyBuffStackCountChangeListeners_30 + 1] = onBuffStackCountChangeListener
end
function BuffUtil.addAnyBuffDestroyListener(self, onBuffDestroyListener)
    local ____Buff__sl_anyBuffDestroyListeners_31 = Buff._sl_anyBuffDestroyListeners
    ____Buff__sl_anyBuffDestroyListeners_31[#____Buff__sl_anyBuffDestroyListeners_31 + 1] = onBuffDestroyListener
end
function BuffUtil.setUiEnable(self, buffTypeId, enable, player)
    if player ~= nil and GetLocalPlayer() ~= player then
        return
    end
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType == nil then
        print("设置未注册的Buff图标:" .. buffTypeId)
        return
    end
    buffType.uiEnable = enable
end
function BuffUtil.setBuffTypeUbertip(self, buffTypeId, ubertip, player)
    if player ~= nil and GetLocalPlayer() ~= player then
        return
    end
    local buffType = ____exports.default:getBuffType(buffTypeId)
    if buffType == nil then
        print("设置未注册的Buff提示:" .. buffTypeId)
        return
    end
    buffType.ubertip = ubertip
end
BuffUtil.buffTypes = {}
return ____exports

File: src/solar/solar-common/util/system/BuffUtil.ts
import Buff from "@/tool/Buff";
import DataBase from "@/common/DataBase";
import BaseUtil from "../BaseUtil";

/**
 * buff = 增益效果
 * 太阳模拟buff系统 工具类
 *
 * (buff不单包括常规的增益减益状态 还包括羁绊、神器、入体装备、境界、法宝、丹药等任何自定义属性附加)
 *
 *
 *   buff会经过4个生命周期
 *    onCreated 创建buff
 *    onUpdate buff更新时 默认为每0.1秒调此方法
 *     onInterval 间隔到期
 *    onDestroy  buff结束销毁时
 */
declare type BuffSet = { //buff集合
    [buffTypeId: string]: Buff
}


/**
 * @deprecated see ActorUtil 推荐使用演员buff 以跟演员系统 结合使用
 */
export default class BuffUtil {
    /**这里也保存一下 以保证遍历顺序跟注册顺序一致*/
    private static buffTypes: AppBuffType[] = []
    /**
     * 注册buff类型
     * @param buffType
     */
    static registerBuffType(buffType: AppBuffType) {
        //这里保存一下保证读取时快速根据ID读取
        if (DataBase.getSolarBuffType(buffType.id) == null) {
            DataBase.setSolarBuffType(buffType.id, buffType);
            BuffUtil.buffTypes.push(buffType)
        } else {
            log.errorWithTraceBack("不能重复注册BuffType:" + buffType.id + " -> " + tostring(buffType.name) + " -> " + DataBase.getSolarBuffType(buffType.id).name)
        }

    }

    /**
     * 注册buff 事件
     * 不能重复注册
     */
    static registerBuffCreatedEvent(buffTypeId: string, onCreated: (buff: Buff) => void) {
        let buffType = BuffUtil.getBuffType(buffTypeId);
        if (buffType.onCreated) {
            log.errorWithTraceBack("不能重复注册Buff事件:" + buffTypeId)
            return
        }
        buffType.onCreated = onCreated;
    }

    /**
     * 注册buff更新事件 取决于buff更新系统 通常为0.1秒更新一次 可用则绑定UI动画
     * @param buffTypeId
     * @param onUpdate
     */
    static registerBuffUpdateEvent(buffTypeId: string, onUpdate: (buff: Buff, tpf: number) => void) {
        let buffType = BuffUtil.getBuffType(buffTypeId);
        if (buffType.onUpdate) {
            log.errorWithTraceBack("不能重复注册Buff事件:" + buffTypeId)
            return
        }
        buffType.onUpdate = onUpdate;
    }

    /**
     * 注册buff 间隔到期事件
     */
    static registerBuffIntervalEvent(buffTypeId: string, onInterval: (buff: Buff) => void) {
        let buffType = BuffUtil.getBuffType(buffTypeId);
        if (buffType.onInterval) {
            log.errorWithTraceBack("不能重复注册Buff事件:" + buffTypeId)
            return
        }
        buffType.onInterval = onInterval;
    }

    /**
     * 注册buff 层数改变事件
     */
    static registerBuffStackCountChangeEvent(buffTypeId: string, onStackCountChange: (buff: Buff, stackAdd: number) => void) {
        let buffType = BuffUtil.getBuffType(buffTypeId);
        if (buffType.onStackCountChange) {
            log.errorWithTraceBack("不能重复注册Buff事件:" + buffTypeId)
            return
        }
        buffType.onStackCountChange = onStackCountChange;
    }

    /**
     * 注册buff 销毁事件
     */
    static registerBuffDestroyEvent(buffTypeId: string, onDestroy: (buff: Buff) => void) {
        let buffType = BuffUtil.getBuffType(buffTypeId);
        if (buffType.onDestroy) {
            log.errorWithTraceBack("不能重复注册Buff事件:" + buffTypeId)
            return
        }
        buffType.onDestroy = onDestroy;
    }


    /**
     * 根据buffTypeId返回Buff类型
     * @param buffTypeId
     */
    static getBuffType(buffTypeId: string): AppBuffType {
        let bt = DataBase.getSolarBuffType(buffTypeId);
        return bt;
    }


    /**
     * 给单位上一个buff
     * @param unit 给哪个单位上
     * @param buffTypeId
     * @param caster buff来源单位
     * @param initBuff 初始化buff回调 会在  buffType.onCreated 之前调用  以给buff设置初始自定义值
     */
    static addBuff(unit: unit, buffTypeId: string, caster?: unit, initBuff?: (buff: Buff) => void): Buff {
        let buffType = DataBase.getSolarBuffType(buffTypeId);
        if (!buffType) {
            log.errorWithTraceBack("buff不存在!请先使用registerBuffType注册Buff类型! id=" + buffTypeId)
            return null;
        }
        //判断是否已存在buff
        let oldUnitBuff = BuffUtil.getUnitBuff(unit, buffTypeId);
        if (oldUnitBuff) {
            if (buffType.stackCountMax && buffType.stackCountMax != 0) {//可叠加
                //-1 = 无限叠加
                if (buffType.stackCountMax == -1 || oldUnitBuff.stackCount < buffType.stackCountMax) {
                    oldUnitBuff.stackCount++;//增加叠加层数
                    if (buffType.stackDurMode) {
                        BaseUtil.runLater(buffType.dur, () => {
                            oldUnitBuff.stackCount--;
                        })
                    }
                }
            }
            oldUnitBuff.refresh()
            return oldUnitBuff;
        }
        //新建buff
        let buff = new Buff(buffType, unit, caster, initBuff);
        return buff;
    }


    /**
     * 获取某个玩家的所有单位拥有的增益效果的 第一个
     * （使用场景：比如获取羁绊UI显示的信息）
     * @param player
     * @param buffTypeId
     */
    static getPlayerUnitBuff(player: player, buffTypeId: string): Buff {
        GroupEnumUnitsOfPlayer(_sl_tempGroup3, player, null)
        for (let i = 0; i <= 1000000; i++) {
            let unitHandle = FirstOfGroup(_sl_tempGroup3)
            if (!IsHandle(unitHandle)) {
                break
            }
            let buff = BuffUtil.getUnitBuff(unitHandle, buffTypeId);
            if (buff?.buffType.id == buffTypeId) {
                return buff
            }
            GroupRemoveUnit(_sl_tempGroup3, unitHandle)
        }
        return null;
    }

    /**
     * 获取单位的某个类型的buff
     * @param unit
     * @param buffTypeId
     */
    static getUnitBuff(unit: unit, buffTypeId: string): Buff {
        return BuffUtil.getUnitBuffs(unit)?.[buffTypeId];
    }

    /**
     * 判断是否拥有某个类型的buff
     * @param unit
     * @param buffTypeId
     */
    static hasUnitBuff(unit: unit, buffTypeId: string): boolean {
        return BuffUtil.getUnitBuffs(unit)?.[buffTypeId] != null;
    }

    /**
     * 获取单位身上所有buff
     * @param unit
     */
    static getUnitBuffs(unit: unit): BuffSet {
        let unitSolarData = DataBase.getUnitSolarData(unit, false);
        return unitSolarData?._SL_solarBuffSet;
    }

    /**
     * 获取单位身上所有buff 的属性值
     * @param unit
     */
    static getUnitBuffAttributes(unit: unit): AppAttribute[] {
        let unitSolarData = DataBase.getUnitSolarData(unit, false);
        let solarBuffSet: BuffSet = unitSolarData?._SL_solarBuffSet;
        if (!solarBuffSet) {
            return null;
        }
        let attributes: AppAttribute[] = [];
        for (let solarBuffSetKey in solarBuffSet) {
            let attribute = solarBuffSet[solarBuffSetKey]?.attribute;
            if (attribute) {
                attributes.push(attribute)
            }
        }
        return attributes;
    }

    /**
     * 遍历所有buff类型
     * @param callback 遍历回调函数
     * @param buffTypeClass 增益效果类型的类别
     */
    static forAllBuffTypes(callback: (buffType: AppBuffType) => void, buffTypeClass?: string): void {
        //
        for (let buffType of BuffUtil.buffTypes) {
            if (buffTypeClass) {
                if (buffTypeClass == buffType.class) {
                    callback(buffType)
                }
            } else {
                callback(buffType)
            }
        }
    }

    /**
     * 遍历地图上所有单位的所有buffs
     * @param callback
     * @param buffTypeId 只对某个buffType做遍历
     */
    static forAllBuffs(callback: (buff: Buff) => void, buffTypeId?: string): void {
        if (buffTypeId) {//遍历单个类型
            //遍历所有
            DataBase.forUnitSolarDatas((id, solarData: AppData) => {
                let _SL_solarBuffSet: BuffSet = solarData?._SL_solarBuffSet;
                if (_SL_solarBuffSet) {
                    for (let tempBuffTypeId in _SL_solarBuffSet) {
                        let buff = _SL_solarBuffSet[tempBuffTypeId];
                        if (buff?.buffType.id == buffTypeId) {
                            callback(buff)
                        }
                    }
                }
            })
        } else {
            //遍历所有
            DataBase.forUnitSolarDatas((id, solarData: AppData) => {
                let _SL_solarBuffSet: BuffSet = solarData?._SL_solarBuffSet;
                if (_SL_solarBuffSet) {
                    for (let tempBuffTypeId in _SL_solarBuffSet) {
                        callback(_SL_solarBuffSet[tempBuffTypeId])
                    }
                }
            })
        }

    }

    /**
     * 遍历此单位的 所有buff
     */
    static forUnitBuffs(unit: unit, callback: (buff: Buff) => void): void {
        let buffSet = BuffUtil.getUnitBuffs(unit);
        if (!buffSet) {
            return
        }
        for (let buffTypeId in buffSet) {
            let buff = buffSet[buffTypeId];
            if (buff) {
                callback(buff)
            }
        }

    }

    /**
     * 清除单位某个类型的buff
     * @param unit
     * @param buffTypeId
     */
    static clearUnitBuff(unit: unit, buffTypeId: string) {
        BuffUtil.getUnitBuff(unit, buffTypeId)?.destroy();
    }

    /**
     * 清除单位所有buff
     * @param unit 单位
     * @param onlyDeBuff 只清除debuff
     */
    static clearUnitBuffs(unit: unit, onlyDeBuff = false) {
        let buffs: BuffSet = BuffUtil.getUnitBuffs(unit);
        if (!buffs) {
            return;
        }
        for (let buffsKey in buffs) {
            if (onlyDeBuff) {
                if (buffs[buffsKey]?.buffType.isDebuff) {
                    buffs[buffsKey].destroy();
                }
            } else {
                buffs[buffsKey]?.destroy();
            }

        }
    }

    /**
     * 清除单位身上某个类别的所有buff
     * @param unit 单位
     * @param clazz
     */
    static clearUnitBuffsByClass(unit: unit, clazz: string) {
        let buffs: BuffSet = BuffUtil.getUnitBuffs(unit);
        if (!buffs) {
            return;
        }
        for (let buffsKey in buffs) {
            let buff = buffs[buffsKey];
            if (buff?.buffType?.class == clazz) {
                buff.destroy()
            }
        }
    }


    /**
     * 添加任意buff创建事件 监听回调
     */
    static addAnyBuffCreatedListener(onBuffCreatedListener: (buff: Buff) => void) {
        Buff._sl_anyBuffCreatedListeners.push(onBuffCreatedListener);
    }

    /**
     * 添加任意buff叠加层数改变事件 监听回调
     */
    static addAnyBuffStackCountChangeListener(onBuffStackCountChangeListener: (buff: Buff, stackAdd: number) => void) {
        Buff._sl_anyBuffStackCountChangeListeners.push(onBuffStackCountChangeListener);
    }

    /**
     * 添加任意buff销毁事件 监听回调
     */
    static addAnyBuffDestroyListener(onBuffDestroyListener: (buff: Buff) => void) {
        Buff._sl_anyBuffDestroyListeners.push(onBuffDestroyListener);
    }


    /**
     *
     * @param buffTypeId 增益效果类型id
     * @param enable 启用状态
     * @param player 改变指定玩家的图标启用状态 不填则给所有玩家修改
     */
    static setUiEnable(buffTypeId: string, enable: boolean, player?: player): void {
        if (player != null && GetLocalPlayer() != player) {//异步给指定的玩家改
            return;
        }
        let buffType = BuffUtil.getBuffType(buffTypeId)
        if (buffType == null) {
            print("设置未注册的Buff图标:" + buffTypeId)
            return;
        }
        buffType.uiEnable = enable;
    }


    /**
     * 可以改变buff类型 的提示信息
     * @param buffTypeId
     * @param ubertip
     * @param player 为指定玩家异步修改
     */
    static setBuffTypeUbertip(buffTypeId: string, ubertip: string, player?: player): void {
        if (player != null && GetLocalPlayer() != player) {//异步给指定的玩家改
            return;
        }
        let buffType = BuffUtil.getBuffType(buffTypeId)
        if (buffType == null) {
            print("设置未注册的Buff提示:" + buffTypeId)
            return;
        }
        buffType.ubertip = ubertip;
    }
}